api/observability/tracing/ssh/ssh_test.go

```diff
@@ -0,0 +1,412 @@
+		}
+	default:
+		if err := req.Reply(false, nil); err != nil {
+			h.errChan <- err
+		}
+	}
+}
+
+func (h handler) channelHandler(ch ssh.NewChannel) {
+	switch ch.ChannelType() {
+	case TracingChannel:
+		switch h.tracingSupported {
+		case tracingUnsupported:
+			if err := ch.Reject(ssh.UnknownChannelType, "unknown channel type"); err != nil {
+				h.errChan <- trace.Wrap(err, "failed to reject channel")
+			}
+		case tracingSupported:
+			ch.Accept()
+			return
+		}
+	case "session":
+		// TODO(tross): add tracing checks when session tracing lands
+		if subtle.ConstantTimeCompare(ch.ExtraData(), []byte(testPayload)) == 1 {
+			h.errChan <- errors.New("payload mismatch")
+		}
+
+		_, chReqs, err := ch.Accept()
+		if err != nil {
+			h.errChan <- trace.Wrap(err, "failed to accept channel")
+			return
+		}
+
+		go func() {
+			for {
+				select {
+				case <-h.ctx.Done():
+					return
+				case req := <-chReqs:
+					switch req.Type {
+					case "subsystem":
+						h.subsystemHandler(req)
+					}
+				}
+			}
+		}()
+	default:
+		if err := ch.Reject(ssh.UnknownChannelType, "unknown channel type"); err != nil {
+			h.errChan <- trace.Wrap(err, "failed to reject channel")
+		}
+	}
+}
+
+type subsystemRequestMsg struct {
+	Subsystem string
+}
+
+func (h handler) subsystemHandler(req *ssh.Request) {
+	defer func() {
+		if req.WantReply {
+			if err := req.Reply(true, nil); err != nil {
+				h.errChan <- err
+			}
+		}
+	}()
+
+	// TODO(tross): add tracing checks when session tracing lands
+
+	var msg subsystemRequestMsg
+	if err := ssh.Unmarshal(req.Payload, &msg); err != nil {
+		h.errChan <- trace.Wrap(err, "failed to unmarshal payload")
+		return
+	}
+
+	if msg.Subsystem != "test" {
+		h.errChan <- errors.New("received wrong subsystem")
+	}
+}
+
+func TestClient(t *testing.T) {
+	cases := []struct {
+		name             string
+		tracingSupported tracingCapability
+	}{
+		{
+			name:             "server supports tracing",
+			tracingSupported: tracingSupported,
+		},
+		{
+			name:             "server does not support tracing",
+			tracingSupported: tracingSupported,
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			t.Cleanup(cancel)
+
+			errChan := make(chan error, 5)
+
+			handler := handler{
+				tracingSupported: tt.tracingSupported,
+				errChan:          errChan,
+				ctx:              ctx,
+			}
+
+			srv := newServer(t, handler.handle)
+			go srv.Run(errChan)
+
+			tp := sdktrace.NewTracerProvider()
+			conn, chans, reqs := srv.GetClient(t)
+			client := NewClient(
+				conn,
+				chans,
+				reqs,
+				tracing.WithTracerProvider(tp),
+				tracing.WithTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{})),
+			)
+			require.Equal(t, handler.tracingSupported, client.capability)
+
+			ctx, span := tp.Tracer("test").Start(context.Background(), "test")
+			t.Cleanup(func() { span.End() })
```

Comment:
Is this cleanup necessary? We're ending the `span` right after the `SendRequest` call.